### PR TITLE
Value set: handle with, array and struct expressions more accurately

### PR DESCRIPTION
**Pull request by smowton**
We may have a suffix indicating that the result of a get_value_set_rec query
would effectively be projected down to particular struct member or array indices
on completion (and so we in fact do this eagerly at the bottom of the recursion,
rather than at the top as one might expect). However, the ID_array, ID_struct and
ID_with expressions are widening -- one or more of their operands is integrated
into a larger type -- so we should consume suffix entries introduced by ID_member
expressions if available.

### CODE REVIEW

File: src/pointer-analysis/value_set.cpp [hunk 1]
```diff
@@ -344,6 +344,30 @@ void value_sett::get_value_set(
   get_value_set_rec(tmp, dest, "", tmp.type(), ns);
 }
 
+/// Check if 'suffix' starts with 'field'.
+/// Suffix is delimited by periods and '[]' (array access) tokens, so we're
+/// looking for ".field($|[]|.)"
+static bool suffix_starts_with_field(
+  const std::string &suffix, const std::string &field)
```
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp:351]: I think each call of this function would like to pass an `irep_idt` as second argument; I'd thus suggest to do the `id2string` within this function. The same is true for `strip_first_field_from_suffix`.

**Review by tautschnig — APPROVED** [review]
1) I think this has huge potential for improving both performance of symbolic execution and precision of slicing -- I'd invite you to try the current `KNOWNBUG` tests of the slicer.
2) Yes, there should of course be tests :-) - my suggested approach to testing is checking the output of --program-only or --show-vcc on some example where more fine-grained dereferencing appears.
3) String manipulation is of course terrible and likely this should be replaced by expressions. But that will lead to a much bigger rewrite I suppose.

**Review by owen-mc-diffblue — APPROVED** [review]
It looks good. Approving assuming you will add tests and get CI passing.

**Comment by smowton** [comment]
@tautschnig since the relationship between particular expressions and C code is nebulous and prone to change I went with unit testing. I also added the Java example that brought the bug to our attention in the first place, though local testing reveals that particular manifestation of the problem is fixed both by this PR and also by #2585, since that eliminates WITH expressions for this case.

**Comment by smowton** [comment]
Re: the slicer tests, I'm afraid as the person with an interest in them I'll leave that experiment to you ;)

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 943d60c8b67194744dc1aec27fdfe7e1c655e412).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80730469

**Comment by smowton** [comment]
@tautschnig @peterschrammel this is now passing and ready for re-review

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
@tautschnig Alas, I'm not sure this will improve either as I think they both use different pointer analyses.

**Pull request merged**